### PR TITLE
Transfer learning

### PR DESCRIPTION
**Pull request by semibran**
- Add data augmentation
- Add early stopping
- Add transfer learning with fine tuning

The tools we have to counterbalance overfit atm are batch size, data augmentation, dropout, fine tuning learning rate, and potentially(?) batch normalization momentum.

generally the settings we have right now work fine up until the fine tuning stage, after which point we see diminishing returns from validation accuracy with great overfit. Here fine tuning starts at around the 22 epoch mark:
history

The above model uses MobileNet v2 96x96 which seems to cap out around 80% validation accuracy, whereas models that accept higher resolutions can exceed 90%

### CODE REVIEW

File: configs/model.py [hunk 1]
```diff
@@ -10,10 +12,15 @@
 
 SEED = 0
 VALIDATION_SPLIT = 0.2
-IMAGE_SIZE = (96, 72)
 
-MODEL_NUM_EPOCHS = 5
+MODEL_FEATURE_EXTRACTOR, IMAGE_SIZE = config_transfer(size=299)
```
Comment (semibran) [inline configs/model.py:16]: `size` corresponds to width/height of a square image.... it's a tunable param with values defined via [`configs/transfer.py`](https://github.com/JonathanPaugh/comp4989-project/pull/7/files#diff-b5d2587c85a01b1d6eed22e5c58660bd128735f531b6b5d2107770e2c9326678R1-R7) -- higher means more accuracy but slower (and more memory-intensive)

For perspective `size=96` results in epoch times of approximately 10s for initial fit and 1m fine tuning, while `size=299` takes 4m/20m per epoch respectively

File: configs/model.py [hunk 2]
```diff
@@ -10,10 +12,15 @@
...
+MODEL_EARLY_STOPPING_PATIENCE = 3  # stop after x consecutive epochs with no improvement
+MODEL_NUM_EPOCHS = 50
+MODEL_FINE_TUNING = True
+MODEL_FINE_TUNING_NUM_EPOCHS = 50
+MODEL_FINE_TUNING_LEARNING_RATE = 1e-5         # defaults to 1e-4
+MODEL_FINE_TUNING_BATCH_NORM_MOMENTUM = 0.997  # defaults to 0.99
```
Comment (semibran) [inline configs/model.py:22]: these two params control how quickly the model learns during fine tuning

Typically you want a low learning rate here since fine tuning will generally overfit the training data quickly. `1e-5` seems pretty standard from what I've seen

I've yet to test if batch norm momentum does anything considering that the batch normalization layers are meant to be frozen into inference mode, but included here for completion's sake

File: core/model.py [hunk 3]
```diff
@@ -1,48 +1,118 @@
...
+    )
+
+    # HACK: define model via functional API to ensure feature extractor runs in inference mode
+    inputs = Input(shape=INPUT_SHAPE)
+    x = data_augmentation(inputs)
+    x = Rescaling(1./255)(x)
+    x = feature_extractor(x, training=False)  # force run in inference mode
+    x = Dropout(config.MODEL_DROPOUT_RATE)(x)
+    outputs = Dense(num_classes, activation='softmax')(x)
+    model = Model(inputs, outputs)
```
Comment (semibran) [inline core/model.py:57]: more details on [inference mode](https://keras.io/guides/transfer_learning/#do-a-round-of-finetuning-of-the-entire-model) -- basically the idea behind fine tuning is that we want to unfreeze some of the layers in the pre-trained model so some of the higher-level features can adapt to suit our problem domain, but if we unfreeze all the layers it'll apparently ruin the weights of [the batch normalization layers](https://keras.io/api/layers/normalization_layers/batch_normalization/) that were previously generalized to the imagenet dataset (small-scale "catastrophic forgetting")

File: configs/transfer.py [hunk 4]
```diff
@@ -0,0 +1,11 @@
+FEATURE_EXTRACTOR_MAP = {
+  299: 'https://tfhub.dev/google/imagenet/inception_v3/feature_vector/5',
+  224: 'https://tfhub.dev/google/imagenet/mobilenet_v2_100_224/feature_vector/5',
+  192: 'https://tfhub.dev/google/imagenet/mobilenet_v2_100_192/feature_vector/5',
+  128: 'https://tfhub.dev/google/imagenet/mobilenet_v2_100_128/feature_vector/5',
+  96: 'https://tfhub.dev/google/imagenet/mobilenet_v2_100_96/feature_vector/5',
+}
```
Comment (semibran) [inline configs/transfer.py:7]: The 100 parameter here indicates a depth multiplier of 1.00; it's essentially another training speed/test accuracy tradeoff where lower has a similar effect to decreasing image resolution, except it's related to number of channels(?) instead of image size. MobileNet distributions ship with depth multipliers ranging from 0.25 to 1.00, with 1.00 being the slowest/most accurate.

As smaller depth multipliers result in a smaller number of trainable params, they may also allow us to fine tune memory usage.

**Review by jonathan-paugh — APPROVED** [review]
Time to deploy 🚀
